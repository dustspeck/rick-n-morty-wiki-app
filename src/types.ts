interface IPageInfo {
  count: number;
  pages: number;
  next: string | null;
  prev: string | null;
}
export interface IPage {
  info: IPageInfo;
  results: any[];
}

interface IOrigin {
  name: string;
  url: string;
}

interface ILocation {
  name: string;
  url: string;
}

export type IStatus = 'Alive' | 'Dead' | 'unknown';
export type IGender = 'Female' | 'Male' | 'Genderless' | 'unknown';

export interface ICharacter {
  id: number;
  name: string;
  status: IStatus;
  species: string;
  type: string;
  gender: IGender;
  origin: IOrigin;
  location: ILocation;
  image: string;
  episode: string[];
  url: string;
  created: string;
}

export interface ICharactersPage extends IPage {
  results: ICharacter[];
}
