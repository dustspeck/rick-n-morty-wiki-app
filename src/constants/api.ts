export const BASE_URL = 'https://rickandmortyapi.com/api';
export const CHARACTERS_ROUTE = `${BASE_URL}/character`;
