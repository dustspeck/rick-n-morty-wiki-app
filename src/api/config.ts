import axios from 'axios';
import {BASE_URL} from '../constants/api';

export const apiClient = axios.create({baseURL: BASE_URL});
