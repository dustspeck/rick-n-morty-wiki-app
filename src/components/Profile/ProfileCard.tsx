import {
  StyleSheet,
  View,
  Dimensions,
  TouchableWithoutFeedback,
  Text,
  Image,
} from 'react-native';
import React, {memo} from 'react';
import {BORDER_RADIUS} from '../../constants/styles';
import {BLACK, CARD_BG, LIGHT_GREY} from '../../constants/colors';
import StatusIndicator from '../common/StatusIndicator';
import {ICharacter} from '../../types';
import GenderIndicator from '../common/GenderIndicator';

const {width} = Dimensions.get('screen');

const ProfileCard = memo(
  ({data}: {data: ICharacter}) => {
    return (
      <TouchableWithoutFeedback>
        <View style={styles.card}>
          <Image source={{uri: data.image}} style={styles.icon} />
          <View style={styles.infoContainer}>
            <View style={styles.info}>
              <Text numberOfLines={1} style={styles.mainText}>
                {data.name}
              </Text>
              <View style={styles.infoRow}>
                <Text
                  numberOfLines={1}
                  style={[styles.subText, {textTransform: 'uppercase'}]}>
                  {data.species}
                </Text>
                <Text numberOfLines={1} style={styles.subText}>
                  {data.type}
                </Text>
              </View>
              <View style={styles.infoRow}>
                <StatusIndicator status={data.status} />
                <GenderIndicator gender={data.gender} />
              </View>
            </View>
            <View style={styles.viewContainer}>
              <Text style={styles.viewText}>â†’</Text>
            </View>
          </View>
        </View>
      </TouchableWithoutFeedback>
    );
  },
  (prev, next) => prev.data.id === next.data.id,
);

export default ProfileCard;

const styles = StyleSheet.create({
  card: {
    display: 'flex',
    flexDirection: 'row',
    width: '100%',
    height: width / 3.2,
    borderRadius: BORDER_RADIUS,
    backgroundColor: CARD_BG,
    marginVertical: 7,
    overflow: 'hidden',
    padding: 5,
  },
  icon: {
    height: '100%',
    width: width / 3.2 - 10,
    borderRadius: BORDER_RADIUS,
  },
  infoContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  info: {
    flex: 1,
    justifyContent: 'space-between',
    paddingRight: 10,
  },
  infoRow: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    overflow: 'hidden',
  },
  mainText: {
    color: BLACK,
    fontSize: 20,
    marginHorizontal: 10,
    marginVertical: 5,
    fontWeight: 'bold',
    textDecorationLine: 'underline',
  },
  subText: {
    color: LIGHT_GREY,
    fontSize: 16,
    marginHorizontal: 10,
    marginVertical: 5,
  },
  viewContainer: {justifyContent: 'center', marginHorizontal: 5},
  viewText: {color: BLACK, fontSize: 22, fontWeight: 'bold'},
});
