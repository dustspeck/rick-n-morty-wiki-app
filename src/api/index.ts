import {AxiosResponse} from 'axios';
import {CHARACTERS_ROUTE} from '../constants/api';
import {ICharactersPage, IEpisode, ILocationAll} from '../types';
import {apiClient} from './config';

/**
 * @description Fetches characters without any pagination.
 * Used for initial characters load.
 * @returns {Promise<ICharactersPage | null>} Returns the first characters page.
 * Returns `null` on errors.
 */
export const getCharacters = async (): Promise<ICharactersPage | null> => {
  try {
    const response: AxiosResponse = await apiClient.get(CHARACTERS_ROUTE);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * @description Fetches characters from the given page url.
 * @param {string} url URL to be fetched from.
 * @returns {Promise<ICharactersPage | null>} Returns the nth characters page.
 * * Returns `null` on errors.
 */
export const getCharactersByURL = async (
  url: string,
): Promise<ICharactersPage | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * Fetches location from the given url.
 * @param { string} url URL to be fetched from.
 * @returns {Promise<ILocationAll | null>} Returns the nth location page.
 * * Returns `null` on errors.
 */
export const getLocationByURL = async (
  url: string,
): Promise<ILocationAll | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ILocationAll = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * Fetches episode from the given url.
 * @param { string} url URL to be fetched from.
 * @returns {Promise<IEpisode | null>} Returns the nth episode.
 * * Returns `null` on errors.
 */
export const getEpisodeByURL = async (
  url: string,
): Promise<IEpisode | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: IEpisode = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};
