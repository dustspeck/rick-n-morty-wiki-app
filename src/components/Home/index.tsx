import React, {useState, useEffect} from 'react';
import {ActivityIndicator} from 'react-native';
import {BLACK} from '../../constants/colors';
import {getCharacters} from '../../services/api/api';
import {ICharacter, IPageInfo} from '../../types';
import SceneBuilder from '../common/SceneBuilder';
import CardsTray from './CardsTray';

const Home = () => {
  const [characters, setCharacters] = useState<ICharacter[]>([]);
  const [latestPageInfo, setLatestPageInfo] = useState<IPageInfo>({
    count: 0,
    pages: 0,
    next: null,
    prev: null,
  });
  const [isLoading, setIsLoading] = useState(true);

  const fetchCharacters = async () => {
    try {
      const page = await getCharacters();
      if (!page) throw 'Error fetching';
      if (!page.results) throw 'No results';
      setCharacters(page.results);
      setLatestPageInfo(page.info);
    } catch (error) {
      console.log(error);
    } finally {
      setIsLoading(false);
    }
  };
  useEffect(() => {
    fetchCharacters();
  }, []);

  return (
    <SceneBuilder>
      {isLoading && <ActivityIndicator color={BLACK} />}
      {!isLoading && (
        <CardsTray
          characters={characters}
          setCharacters={setCharacters}
          latestPageInfo={latestPageInfo}
          setLatestPageInfo={setLatestPageInfo}
        />
      )}
    </SceneBuilder>
  );
};

export default Home;
