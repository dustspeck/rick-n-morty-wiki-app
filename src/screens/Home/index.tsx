import React, {useState, useEffect} from 'react';
import {ActivityIndicator} from 'react-native';
import {BLACK} from '../../constants/colors';
import {getCharacters} from '../../api';
import {ICharacter, IPageInfo} from '../../types';
import SceneBuilder from '../../components/common/SceneBuilder';
import CardsTray from '../../components/Home/CardsTray';

const Home = () => {
  const initPageInfo: IPageInfo = {count: 0, pages: 0, next: null, prev: null};

  const [characters, setCharacters] = useState<ICharacter[]>([]);
  const [latestPageInfo, setLatestPageInfo] = useState<IPageInfo>(initPageInfo);
  const [isLoading, setIsLoading] = useState(true);

  const fetchCharacters = async () => {
    try {
      const page = await getCharacters();
      if (!page) throw 'Error fetching';
      if (!page.results) throw 'No results';
      setCharacters(page.results);
      setLatestPageInfo(page.info);
    } catch (error) {
      console.log(error);
    } finally {
      setIsLoading(false);
    }
  };
  useEffect(() => {
    fetchCharacters(); //initial page load
  }, []);

  return (
    <SceneBuilder>
      {isLoading && <ActivityIndicator testID="Home.Loader" color={BLACK} />}
      {!isLoading && (
        <CardsTray
          characters={characters}
          setCharacters={setCharacters}
          latestPageInfo={latestPageInfo}
          setLatestPageInfo={setLatestPageInfo}
        />
      )}
    </SceneBuilder>
  );
};

export default Home;
