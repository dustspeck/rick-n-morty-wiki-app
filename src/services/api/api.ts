import axios, {AxiosResponse} from 'axios';
import {CHARACTERS_ROUTE} from '../../constants/api';
import {ICharactersPage, ILocationAll} from '../../types';
import {apiClient} from './config';

/**
 * @description Fetches characters without any pagination.
 * Used for initial characters load.
 * @returns {Promise<ICharactersPage | null>} Returns the first characters page.
 * Returns `null` on errors.
 */
export const getCharacters = async (): Promise<ICharactersPage | null> => {
  try {
    const response: AxiosResponse = await apiClient.get(CHARACTERS_ROUTE);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * @description Fetches characters of the given page number.
 * @param {number} page Page number to be fetched.
 * @returns {Promise<ICharactersPage | null>} Returns the nth characters page.
 * * Returns `null` on errors.
 */
export const getCharactersByPage = async (
  page: number,
): Promise<ICharactersPage | null> => {
  try {
    const response: AxiosResponse = await apiClient.get(
      `${CHARACTERS_ROUTE}/page=${page}`,
    );
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * @description Fetches characters from the given page url.
 * @param {string} url URL to be fetched from.
 * @returns {Promise<ICharactersPage | null>} Returns the nth characters page.
 * * Returns `null` on errors.
 */
export const getCharactersByURL = async (
  url: string,
): Promise<ICharactersPage | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};
/**
 * Fetches location from the given url.
 * @param { string} url URL to be fetched from.
 * @returns {Promise<ILocationAll | null>} Returns the nth location page.
 * * Returns `null` on errors.
 */
export const getLocationByURL = async (
  url: string,
): Promise<ILocationAll | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ILocationAll = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};
