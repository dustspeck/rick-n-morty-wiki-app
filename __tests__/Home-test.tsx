import {fireEvent, render, waitFor} from '@testing-library/react-native';

import {ICharactersPage} from '../src/types';
import {getCharacters} from '../src/api';
import Home from '../src/components/Home';
import ProfileCard from '../src/components/Profile/ProfileCard';

const characterPageResponse = {
  info: {count: 0, next: null, pages: 0, prev: null},
  results: [
    {
      id: 20,
      name: 'Ants in my Eyes Johnson',
      status: 'unknown',
      species: 'Human',
      type: 'Human with ants in his eyes',
      gender: 'Male',
      origin: {
        name: 'unknown',
        url: '',
      },
      location: {
        name: 'Interdimensional Cable',
        url: 'https://rickandmortyapi.com/api/location/6',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/20.jpeg',
      episode: ['https://rickandmortyapi.com/api/episode/8'],
      url: 'https://rickandmortyapi.com/api/character/20',
      created: '2017-11-04T22:34:53.659Z',
    },
  ],
} as ICharactersPage;

const mockedNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({navigate: mockedNavigate}),
}));

jest.mock('../src/api');

it('Home renders the profile card', async () => {
  const mGetCharacters = jest.mocked(getCharacters);
  mGetCharacters.mockResolvedValue(characterPageResponse);
  const home = render(<Home />);
  expect(getCharacters).toHaveBeenCalledTimes(1);
  await waitFor(() => home.getByTestId('ProfileCard.Pressable'));
  expect(home).toMatchSnapshot();
});

it('Profile card navigates to profile', () => {
  const data = characterPageResponse.results[0];
  const profileCard = render(<ProfileCard data={data} />);
  fireEvent.press(profileCard.getByTestId('ProfileCard.Pressable'));
  expect(mockedNavigate).toHaveBeenCalledWith('Profile', {data});
});
