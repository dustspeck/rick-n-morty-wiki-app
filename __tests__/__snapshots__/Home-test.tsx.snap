// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders the profile card 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F0EBE3",
        "minHeight": 1334,
        "padding": 20,
      },
      undefined,
      undefined,
    ]
  }
>
  <View>
    <View
      style={
        {
          "minHeight": 1334,
        }
      }
    >
      <RCTScrollView
        ListFooterComponent={
          <Memo
            hasListEnded={false}
          />
        }
        ListHeaderComponent={[Function]}
        data={
          [
            {
              "created": "2017-11-04T22:34:53.659Z",
              "episode": [
                "https://rickandmortyapi.com/api/episode/8",
              ],
              "gender": "Male",
              "id": 20,
              "image": "https://rickandmortyapi.com/api/character/avatar/20.jpeg",
              "location": {
                "name": "Interdimensional Cable",
                "url": "https://rickandmortyapi.com/api/location/6",
              },
              "name": "Ants in my Eyes Johnson",
              "origin": {
                "name": "unknown",
                "url": "",
              },
              "species": "Human",
              "status": "unknown",
              "type": "Human with ants in his eyes",
              "url": "https://rickandmortyapi.com/api/character/20",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        initialNumToRender={10}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={1}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={true}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
        windowSize={1334}
      >
        <View>
          <View
            onLayout={[Function]}
          >
            <View>
              <Text
                style={
                  {
                    "backgroundColor": "#F0EBE3",
                    "color": "#2C3333",
                    "fontSize": 50,
                    "fontWeight": "700",
                  }
                }
              >
                The 
                <Text
                  style={
                    [
                      {
                        "backgroundColor": "#F0EBE3",
                        "color": "#2C3333",
                        "fontSize": 50,
                        "fontWeight": "700",
                      },
                      {
                        "textDecorationLine": "underline",
                      },
                    ]
                  }
                >
                  Rick and Morty
                </Text>
                 Wiki.
              </Text>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="ProfileCard.Pressable"
            >
              <View
                style={
                  {
                    "backgroundColor": "#E4DCCF",
                    "borderRadius": 14,
                    "display": "flex",
                    "flexDirection": "row",
                    "height": 234.375,
                    "marginVertical": 7,
                    "overflow": "hidden",
                    "padding": 5,
                    "width": "100%",
                  }
                }
              >
                <Image
                  source={
                    {
                      "uri": "https://rickandmortyapi.com/api/character/avatar/20.jpeg",
                    }
                  }
                  style={
                    {
                      "borderRadius": 14,
                      "height": "100%",
                      "width": 224.375,
                    }
                  }
                />
                <View
                  style={
                    {
                      "flex": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "justifyContent": "space-between",
                        "paddingRight": 10,
                      }
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        {
                          "color": "#2C3333",
                          "fontSize": 20,
                          "fontWeight": "bold",
                          "marginHorizontal": 10,
                          "marginVertical": 5,
                          "textDecorationLine": "underline",
                        }
                      }
                    >
                      Ants in my Eyes Johnson
                    </Text>
                    <View
                      style={
                        {
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                          "overflow": "hidden",
                        }
                      }
                    >
                      <Text
                        numberOfLines={1}
                        style={
                          [
                            {
                              "color": "#5E6666",
                              "fontSize": 16,
                              "marginHorizontal": 10,
                              "marginVertical": 5,
                            },
                            {
                              "textTransform": "uppercase",
                            },
                          ]
                        }
                      >
                        Human
                      </Text>
                      <Text
                        numberOfLines={1}
                        style={
                          {
                            "color": "#5E6666",
                            "fontSize": 16,
                            "marginHorizontal": 10,
                            "marginVertical": 5,
                          }
                        }
                      >
                        Human with ants in his eyes
                      </Text>
                    </View>
                    <View
                      style={
                        {
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                          "overflow": "hidden",
                        }
                      }
                    >
                      <View
                        style={
                          {
                            "alignItems": "center",
                            "flexDirection": "row",
                          }
                        }
                      >
                        <View
                          style={
                            [
                              {
                                "borderRadius": 5,
                                "height": 5,
                                "marginLeft": 10,
                                "marginRight": 5,
                                "width": 5,
                              },
                              {
                                "backgroundColor": "#476072",
                              },
                            ]
                          }
                        />
                        <Text
                          style={
                            {
                              "color": "#5E6666",
                              "fontSize": 16,
                              "marginVertical": 5,
                            }
                          }
                        >
                          unknown
                        </Text>
                      </View>
                      <View
                        style={
                          {
                            "alignItems": "center",
                            "flexDirection": "row",
                          }
                        }
                      >
                        <Text
                          style={
                            {
                              "color": "#5E6666",
                              "fontSize": 16,
                              "marginVertical": 5,
                            }
                          }
                        >
                          <Text
                            style={
                              {
                                "fontSize": 20,
                              }
                            }
                          >
                            â™‚ 
                          </Text>
                          Male
                        </Text>
                      </View>
                    </View>
                  </View>
                  <View
                    style={
                      {
                        "justifyContent": "center",
                        "marginHorizontal": 5,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#2C3333",
                          "fontSize": 22,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      â†’
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
          >
            <View
              style={
                {
                  "height": 266.8,
                }
              }
            >
              <ActivityIndicator
                color="#476072"
                size="small"
              />
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
